modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/ReadWriteTransactionImpl.java

```diff
@@ -126,10 +135,14 @@ protected CompletableFuture<Void> finish(boolean commit) {
                                         id()
                                 );
                             } else {
-                                return CompletableFuture.completedFuture(null);
+                                return completedFuture(null);
                             }
                         }
                 );
+
+        mainFinishFut.handle((res, e) -> finishFut.get().complete(null));
+
+        return mainFinishFut;
```

Comment:
What's the point of returning mainFinishFut and not finishFut here?

Reply:
We want to get an original future on the first invocation. It can be completed exceptionally.